ARROW-17407: [Doc][FlightRPC] Flight/gRPC best practices (#13873)
apache/arrow: docs/source/cpp/flight.rst

Round robin balancing can cause every client to have an open connection to
every server causing an unexpected number of open connections and a depletion
of resources.

Debugging
---------

Use netstat to see the number of open connections.
For debug - env GRPC_VERBOSITY=info GRPC_TRACE=http will print the initial
headers (on both sides) so you can see if grpc established the connection or
not. It will also print when a message is sent, so you can tell if the
connection is open or not.
Note: "connect" isn't really a connect and we’ve observed that gRPC does not
give you the actual error until you first try to make a call. This can cause
error being reported at unexpected times.

Use ListFlights sparingly
-------------------------

ListFlights endpoint is largely just implemented as a normal GRPC stream
endpoint and can hit transfer bottlenecks if used too much. To estimate data
transfer bottleneck:
5k schemas will serialize to about 1-5 MB/call. Assuming a gRPC localhost
bottleneck of 3GB/s you can at best serve 600-3000 clients/s.

ARROW-15764_ proposes a caching optimisation for server side, but it was not
yet implemented.

Memory cache client-side
------------------------

> Flight uses gRPC allocator wherever possible.

Reviewer: ```suggestion
Flight tries to reuse allocations made by gRPC to avoid redundant
data copies.  However, this means that those allocations may not
be tracked by the Arrow memory pool, and that memory usage behavior,
such as whether free memory is returned to the system, is dependent
on the allocator that gRPC uses (usually the system allocator).
```

fail the request; hence, they can be used to implement custom
authentication methods.

.. _flight-best-practices:

> Best practices

Reviewer: Link to the grpc best practices doc as well?

Author: They are linked below as `best gRPC practices`

==============

gRPC
----

When using default gRPC transport options can be passed to it via

> :member:`arrow::flight::FlightClientOptions::generic_options`. For example:

Reviewer: This is less a best practice, more of a "howto" so I think this should go in the cookbook

.. code-block:: cpp

   auto options = FlightClientOptions::Defaults();
   // Set a very low limit at the gRPC layer to fail all calls
   options.generic_options.emplace_back(GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH, 4);

Also see `best gRPC practices`_ and available `gRPC keys`_.

Re-use clients whenever possible
--------------------------------

Closing clients causes gRPC to close and clean up connections which can take
several seconds per connection. This will stall server and client threads if
done too frequently. Client reuse will avoid this issue.

Don’t round-robin load balance
------------------------------

> Round robin balancing can cause every client to have an open connection to

Reviewer: Link to the gRPC page describing load balancing? https://github.com/grpc/grpc/blob/master/doc/load-balancing.md#round_robin

Author: Added.

> Round robin balancing can cause every client to have an open connection to
> every server causing an unexpected number of open connections and a depletion
> of resources.

Reviewer: ```suggestion
Round robin balancing means every client can have an open connection to
every server, causing an unexpected number of open connections and depleting
server resources.
```

Debugging
---------

Use netstat to see the number of open connections.
For debug - env GRPC_VERBOSITY=info GRPC_TRACE=http will print the initial
headers (on both sides) so you can see if grpc established the connection or
not. It will also print when a message is sent, so you can tell if the
connection is open or not.

> Note: "connect" isn't really a connect and we’ve observed that gRPC does not
> give you the actual error until you first try to make a call. This can cause
> error being reported at unexpected times.

Reviewer: ```suggestion
gRPC may not report connection errors until a call is actually made.
Hence, to detect connection errors when creating a client, some sort
of dummy RPC should be made.
```

> Use ListFlights sparingly

Reviewer: This is too specific to have, IMO

Author: Removed from here and from Go doc.

-------------------------

ListFlights endpoint is largely just implemented as a normal GRPC stream
endpoint and can hit transfer bottlenecks if used too much. To estimate data
transfer bottleneck:
5k schemas will serialize to about 1-5 MB/call. Assuming a gRPC localhost
bottleneck of 3GB/s you can at best serve 600-3000 clients/s.

ARROW-15764_ proposes a caching optimisation for server side, but it was not
yet implemented.

> Memory cache client-side

Reviewer: this isn't specific to the client

Author: Changed.

------------------------

Flight uses gRPC allocator wherever possible.

> gRPC will spawn an unbounded number of threads for concurrent clients. Those

Reviewer: Move this to 'excessive traffic' below?

Author: Done I think. Diff is confusing me right now.

threads are not necessarily cleaned up (cached thread pool in java parlance).
glibc malloc clears some per thread state and the default tuning never clears
caches in some workloads. But you can explicitly tell malloc to dump caches.
See ARROW-16697_ as an example.

> A quick way of testing: attach to the process with a debugger and call malloc_trim

Reviewer: or call ReleaseUnused on the system pool

Author: Added.

Excessive traffic
-----------------

There are basically two ways to handle excessive traffic:
* unbounded thread pool -> everyone gets serviced, but it might take forever.
This is what you are seeing now.
bounded thread pool -> Reject connections / requests when under load, and have
clients retry with backoff. This also gives an opportunity to retry with a

> different node. Not everyone gets serviced but quality of service stays consistent.

Reviewer: we should add code to configure these modes

Author: You mean like setting `GRPC_ARG_MIN_RECONNECT_BACKOFF_MS` to `200ms` for bounded and `0ms` for unbounded? Perhaps another setting I'm not aware of?

Author: Added `GRPC_ARG_MIN_RECONNECT_BACKOFF_MS` example.

> 

Reviewer 2: ```suggestion

Limiting DoPut Batch Size
--------------------------

You may wish to limit the maximum size a client can submit to a server through
DoPut, to prevent a request from taking up too much memory on the server. On
the client-side, set `write_size_limit_bytes` on `FlightClientOptions`. On the 
server-side, set the gRPC option `GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH`.
The client-side option will return an error that can be retried with smaller batches,
while the server-side limit will close out the connection. Setting both can be wise, since
the former provides a better user experience but the latter may be necessary to defend
against impolite clients.

```

Closing unresponsive connections
--------------------------------

1. A stale connection can be closed using
   :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the

>    stop token at connection establishment time.

Reviewer: this needs code samples

(it would be great if all code samples linked to the cookbook)

Author: Perhaps we could add samples for now and they would be moved out as cookbooks become available?
Also I don't see an issue if we duplicate atomic pieces of the cookbooks here.

Author: Yeah, agreed.

Reviewer: We can maybe merge this then add issues for moving each of these examples to the cookbook.

fail the request; hence, they can be used to implement custom
authentication methods.

.. _flight-best-practices:

Best practices
==============

gRPC
----

When using default gRPC transport options can be passed to it via
:member:`arrow::flight::FlightClientOptions::generic_options`. For example:

.. tab-set::

   .. tab-item:: C++

      .. code-block:: cpp

>          auto options = FlightClientOptions::Defaults();

Author: Changed. Please check if it makes sense.

         // Set a very low limit at the gRPC layer to fail all calls
         options.generic_options.emplace_back(GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH, 4);

   .. tab-item:: Python

      .. code-block:: cpp

         // Set a very low limit at the gRPC layer to fail all calls
         generic_options = [("GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH", 4)]
         client = pyarrow.flight.FlightClient(server_uri, generic_options=generic_options)

Also see `best gRPC practices`_ and available `gRPC keys`_.

Re-use clients whenever possible
--------------------------------

> Closing clients causes gRPC to close and clean up connections which can take

Author: I'm not sure what's more costly time wise - creating a client or closing it. I don't know that somebody measured this. Either way it's remedied by re-using clients.

> Closing clients causes gRPC to close and clean up connections which can take

Reviewer: Maybe word it more generically, ~"Creating and closing clients requires setup/teardown on the client and server which can take away from actually handling RPCs. Reuse clients whenever possible to avoid this. Note that clients are thread-safe."

several seconds per connection. This will stall server and client threads if
done too frequently. Client reuse will avoid this issue.

Don’t round-robin load balance
------------------------------

`Round robin balancing`_ means every client can have an open connection to
every server, causing an unexpected number of open connections and depleting
server resources.

Debugging
---------

> Use netstat to see the number of open connections.

Author: Added.

For debug - env GRPC_VERBOSITY=info GRPC_TRACE=http will print the initial
headers (on both sides) so you can see if grpc established the connection or
not. It will also print when a message is sent, so you can tell if the
connection is open or not.
gRPC may not report connection errors until a call is actually made.
Hence, to detect connection errors when creating a client, some sort
of dummy RPC should be made.

> Memory cache management

Reviewer: ```suggestion
Memory management
```

-----------------------

Flight tries to reuse allocations made by gRPC to avoid redundant
data copies.  However, this means that those allocations may not
be tracked by the Arrow memory pool, and that memory usage behavior,
such as whether free memory is returned to the system, is dependent
on the allocator that gRPC uses (usually the system allocator).

> A quick way of testing: attach to the process with a debugger and call malloc_trim
> or call ReleaseUnused on the system pool.

Reviewer: ```suggestion
A quick way of testing: attach to the process with a debugger and call
malloc_trim, or call ReleaseUnused on the system pool. If memory usage
drops, then likely, there is memory allocated by gRPC or by the
application that the system allocator was holding on to. This can be
adjusted in platform-specific ways; see an investigation in JIRA for
an example of how this works on Linux/glibc. 
```

and link the JIRA

Excessive traffic
-----------------

> gRPC will spawn an unbounded number of threads for concurrent clients. Those

Reviewer: This might not actually be true? e.g. `https://github.com/grpc/grpc/issues/14492`

Ah, it's not actually _unbounded_, just a very large number: https://github.com/grpc/proposal/blob/master/L30-cpp-control-max-threads-in-SyncServer.md

I think the threads are cleaned up now? So the next sentence may not be right

> gRPC will spawn an unbounded number of threads for concurrent clients. Those

Author: Updated. Please see if it makes sense.

threads are not necessarily cleaned up (cached thread pool in java parlance).
glibc malloc clears some per thread state and the default tuning never clears
caches in some workloads. But you can explicitly tell malloc to dump caches.

> See ARROW-16697_ as an example.

Reviewer: This might belong better in the discussion of memory management below

Author: Moved.

There are basically two ways to handle excessive traffic:
* unbounded thread pool -> everyone gets serviced, but it might take forever.
* bounded thread pool -> Reject connections / requests when under load, and have
clients retry with backoff. This also gives an opportunity to retry with a

> different node. Not everyone gets serviced but quality of service stays consistent.

Reviewer: I'd probably word this like ~gRPC's default behavior allows one server to accept many connections from many different clients, but if requests do a lot of work (as they may under Flight), the server may not be able to keep up. Configuring the server to limit the number of clients and reject requests more proactively, and configuring clients to retry with backoff (and potentially connect to a different node), would give more consistent quality of service.

Author: Switched to proposed text.

Limiting DoPut Batch Size
--------------------------

You may wish to limit the maximum size a client can submit to a server through
DoPut, to prevent a request from taking up too much memory on the server. On

> the client-side, set `write_size_limit_bytes` on `FlightClientOptions`. On the 

Author: Added.

server-side, set the gRPC option `GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH`.
The client-side option will return an error that can be retried with smaller batches,
while the server-side limit will close out the connection. Setting both can be wise, since
the former provides a better user experience but the latter may be necessary to defend
against impolite clients.

Closing unresponsive connections
--------------------------------

1. A stale connection can be closed using
   :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the
   stop token at connection establishment time.

   .. tab-set::

      .. tab-item:: C++

         .. code-block:: cpp

      .. tab-item:: Java

         .. code-block:: java

      .. tab-item:: Python

         .. code-block:: python

> 2. Use client timeout.

Reviewer: ```suggestion
2. Use call timeouts. (This is a general gRPC best practice.)
```

   .. tab-set::

      .. tab-item:: C++

         .. code-block:: cpp

      .. tab-item:: Java

         .. code-block:: java

      .. tab-item:: Python

         .. code-block:: python

> 3. There is a long standing ticket for a per-write/per-read timeout instead of a per

Reviewer: This first sentence might be better as a `.. note` since it's not actually actionable

   call timeout (ARROW-6062_), but this is not (easily) possible to implement with the

>    blocking gRPC API. For now one can also do something like set up a background thread

Reviewer: This sort of thing should definitely be a recipe.

> 3. There is a long standing ticket for a per-write/per-read timeout instead of a per
>    call timeout (ARROW-6062_), but this is not (easily) possible to implement with the
>    blocking gRPC API. For now one can also do something like set up a background thread
>    that calls cancel() on a timer and have the main thread reset the timer every time a
>    write operation completes successfully (that means one needs to use to_batches() +
>    write_batch and not write_table).

Reviewer: ```suggestion
3. Client timeouts are not great for long-running streaming calls, where it may
   be hard to choose a timeout for the entire operation. Instead, what is often
   desired is a per-read or per-write timeout so that the operation fails if it
   isn't making progress. This can be implemented with a background thread that
   calls cancel() on a timer, with the main thread resetting the timer every time
   an operation completes successfully. For a fully-worked out example, see the
   Cookbook.
   
   .. note:: There is a long standing ticket for a per-write/per-read timeout
             instead of a per call timeout (ARROW-6062_), but this is not (easily)
             possible to implement with the blocking gRPC API. 
```

fail the request; hence, they can be used to implement custom
authentication methods.

.. _flight-best-practices:

Best practices
==============

gRPC
----

> When using default gRPC transport options can be passed to it via

Reviewer: ```suggestion
When using the default gRPC transport, options can be passed to it via
```

:member:`arrow::flight::FlightClientOptions::generic_options`. For example:

.. tab-set::

   .. tab-item:: C++

      .. code-block:: cpp

         auto options = FlightClientOptions::Defaults();
         // Set the period after which a keepalive ping is sent on transport.
         options.generic_options.emplace_back(GRPC_ARG_KEEPALIVE_TIME_MS, 60000);

   .. tab-item:: Python

      .. code-block:: python

         # Set the period after which a keepalive ping is sent on transport.
         generic_options = [("GRPC_ARG_KEEPALIVE_TIME_MS", 60000)]
         client = pyarrow.flight.FlightClient(server_uri, generic_options=generic_options)

Also see `best gRPC practices`_ and available `gRPC keys`_.

Re-use clients whenever possible
--------------------------------

Creating and closing clients requires setup and teardown on the client and
server side which can take away from actually handling RPCs. Reuse clients

> whenever possible to avoid this. Note that clients are thread-safe.

Reviewer: ```suggestion
whenever possible to avoid this. Note that clients are thread-safe, so a
single client can be shared across multiple threads.
```

              options.stop_token = stop_source.token();
              stop_source.RequestStop(Status::Cancelled("StopSource"));
              flight_client->DoAction(options, {});

2. Use call timeouts. (This is a general gRPC best practice.)

   .. tab-set::

      .. tab-item:: C++

         .. code-block:: cpp

            FlightCallOptions options;
            options.timeout = TimeoutDuration{0.2};
            Status status = client->GetFlightInfo(options, FlightDescriptor{}).status();

      .. tab-item:: Java

         .. code-block:: java

            Iterator<Result> results = client.doAction(new Action("hang"), CallOptions.timeout(0.2, TimeUnit.SECONDS));

      .. tab-item:: Python

         .. code-block:: python

            options = pyarrow.flight.FlightCallOptions(timeout=0.2)
            result = client.do_action(action, options=options)

3. Client timeouts are not great for long-running streaming calls, where it may
   be hard to choose a timeout for the entire operation. Instead, what is often
   desired is a per-read or per-write timeout so that the operation fails if it
   isn't making progress. This can be implemented with a background thread that

>    calls cancel() on a timer, with the main thread resetting the timer every time

Reviewer: ```suggestion
   calls Cancel() on a timer, with the main thread resetting the timer every time
```

.. tab-set::

   .. tab-item:: C++

      .. code-block:: cpp

         auto options = FlightClientOptions::Defaults();
         // Set the period after which a keepalive ping is sent on transport.
         options.generic_options.emplace_back(GRPC_ARG_KEEPALIVE_TIME_MS, 60000);

   .. tab-item:: Python

      .. code-block:: python

         # Set the period after which a keepalive ping is sent on transport.
         generic_options = [("GRPC_ARG_KEEPALIVE_TIME_MS", 60000)]
         client = pyarrow.flight.FlightClient(server_uri, generic_options=generic_options)

Also see `best gRPC practices`_ and available `gRPC keys`_.

Re-use clients whenever possible
--------------------------------

Creating and closing clients requires setup and teardown on the client and
server side which can take away from actually handling RPCs. Reuse clients
whenever possible to avoid this. Note that clients are thread-safe.

Don’t round-robin load balance
------------------------------

`Round robin load balancing`_ means every client can have an open connection to
every server, causing an unexpected number of open connections and depleting
server resources.

> Debugging disconnects

Reviewer: Maybe just change this to "Debugging connection issues"?

Author: Done.

---------------------

When facing unexpected disconnects on long running connections use netstat to
monitor the number of open connections. If number of connections is much
greater than the number of clients it might cause issues.

For debugging, certain environment variables enable logging in gRPC. For
example, ``env GRPC_VERBOSITY=info GRPC_TRACE=http`` will print the initial
headers (on both sides) so you can see if gRPC established the connection or
not. It will also print when a message is sent, so you can tell if the
connection is open or not.

gRPC may not report connection errors until a call is actually made.
Hence, to detect connection errors when creating a client, some sort
of dummy RPC should be made.

Memory management
-----------------------

Flight tries to reuse allocations made by gRPC to avoid redundant
data copies. However, this means that those allocations may not
be tracked by the Arrow memory pool, and that memory usage behavior,
such as whether free memory is returned to the system, is dependent
on the allocator that gRPC uses (usually the system allocator).

A quick way of testing: attach to the process with a debugger and call

> malloc_trim, or call :func:`ReleaseUnused <arrow::MemoryPool::ReleaseUnused>`

Reviewer: ```suggestion
``malloc_trim``, or call :func:`ReleaseUnused <arrow::MemoryPool::ReleaseUnused>`
```

on the system pool. If memory usage drops, then likely, there is memory
allocated by gRPC or by the application that the system allocator was holding
on to. This can be adjusted in platform-specific ways; see an investigation

> in JIRA for an example of how this works on Linux/glibc.

Reviewer: Which jira?

Author: We were referring to the same on twice I think. Changed.

> malloc can be explicitly told to dump caches. See ARROW-16697_ as an example.

Reviewer: glibc malloc?

Excessive traffic
-----------------

gRPC will spawn up to max threads quota of threads for concurrent clients. Those
threads are not necessarily cleaned up (a "cached thread pool" in Java parlance).
glibc malloc clears some per thread state and the default tuning never clears
caches in some workloads.

gRPC's default behavior allows one server to accept many connections from many
different clients, but if requests do a lot of work (as they may under Flight),
the server may not be able to keep up. Configuring the server to limit the number

> of clients and reject requests more proactively, and configuring clients to retry

Reviewer: How do you configure the server to do this?

Reviewer: (IIRC, it's not possible without modifying Flight…)

Author: I think this is referring to gRPC settings.

Reviewer: I would say either remove this text or have a code sample of how to set the server-side limit

Author: [See here](https://github.com/apache/arrow/pull/13873#discussion_r972262181)

> gRPC's default behavior allows one server to accept many connections from many
> different clients, but if requests do a lot of work (as they may under Flight),
> the server may not be able to keep up. Configuring the server to limit the number
> of clients and reject requests more proactively, and configuring clients to retry

Author: How about this:
```suggestion
gRPC's default behaviour allows one server to accept many connections from many
different clients, but if requests do a lot of work (as they may under Flight),
the server may not be able to keep up. Configuring clients to retry
```

with backoff (and potentially connect to a different node), would give more
consistent quality of service.

.. tab-set::

   .. tab-item:: C++

      .. code-block:: cpp

         auto options = FlightClientOptions::Defaults();
         // Set the minimum time between subsequent connection attempts.
         options.generic_options.emplace_back(GRPC_ARG_MIN_RECONNECT_BACKOFF_MS, 2000);

   .. tab-item:: Python

      .. code-block:: python

         # Set the minimum time between subsequent connection attempts.
         generic_options = [("GRPC_ARG_MIN_RECONNECT_BACKOFF_MS", 2000)]
         client = pyarrow.flight.FlightClient(server_uri, generic_options=generic_options)

Limiting DoPut Batch Size
--------------------------

> You may wish to limit the maximum size a client can submit to a server through

Reviewer: ```suggestion
You may wish to limit the maximum batch size a client can submit to a server through
```

DoPut, to prevent a request from taking up too much memory on the server. On
the client-side, set :member:`arrow::flight::FlightClientOptions::write_size_limit_bytes`.
On the server-side, set the gRPC option ``GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH``.
The client-side option will return an error that can be retried with smaller batches,
while the server-side limit will close out the connection. Setting both can be wise, since
the former provides a better user experience but the latter may be necessary to defend
against impolite clients.

Closing unresponsive connections
--------------------------------

> 1. A stale connection can be closed using

Reviewer: This is more a way to stop a long-running RPC call, than a connection specifically

Reviewer: Actually all of these are so the section should just be renamed, and remove/reword any text referring to connections

Author: Done.

>    :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the

Reviewer: This is FlightCallOptions not ClientOptions

Author: I think this is already `FlightCallOptions` just the diff is off.

Reviewer: I still see ClientOptions here

Author: Saw it here now too. Fixed.

>    :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the

Author: ```suggestion
   :member:`arrow::flight::FlightCallOptions::stop_token`. This requires recording the
```
